src/Features/CSharp/Portable/UseUTF8StringLiteral/UseUTF8StringLiteralRefactoringProvider.cs

```diff
@@ -0,0 +1,192 @@
+            {
+                // Need to call a method to do the actual rune decoding as it uses stackalloc, and stackalloc
+                // in a loop is a bad idea
+                if (!TryGetNextRune(arrayCreationElements, i, out var rune, out var bytesConsumed))
+                    return false;
+
+                i += bytesConsumed;
+
+                if (builder is not null)
+                {
+                    if (rune.TryGetEscapeCharacter(out var escapeChar))
+                    {
+                        builder.Append('\\');
+                        builder.Append(escapeChar);
+                    }
+                    else
+                    {
+                        builder.Append(rune.ToString());
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private static bool TryGetNextRune(ImmutableArray<IOperation> arrayCreationElements, int startIndex, out Rune rune, out int bytesConsumed)
+        {
+            rune = default;
+            bytesConsumed = 0;
+
+            // We only need max 4 elements for a single Rune
+            var length = Math.Min(arrayCreationElements.Length - startIndex, 4);
+
+            Span<byte> array = stackalloc byte[length];
+            for (var i = 0; i < length; i++)
+            {
+                var element = arrayCreationElements[startIndex + i];
+
+                // First basic check is that the array element is actually a byte
+                if (element.ConstantValue.Value is not byte b)
+                    return false;
+
+                array[i] = b;
+            }
+
+            // If we can't decode a rune from the array then it can't be represented as a string
+            return Rune.DecodeFromUtf8(array, out rune, out bytesConsumed) == System.Buffers.OperationStatus.Done;
+        }
+
+        private static async Task<Document> ConvertToUTF8StringLiteralAsync(Document document, SyntaxNode node, ImmutableArray<IOperation> arrayCreationElements, CancellationToken cancellationToken)
+        {
+            // Get our list of bytes from the array elements
+            using var _ = PooledStringBuilder.GetInstance(out var builder);
+            builder.Capacity = arrayCreationElements.Length;
+            if (!TryConvertToUTF8String(builder, arrayCreationElements))
+            {
+                // We shouldn't get here, because the code fix shouldn't ask for a string value
+                // if the analyzer couldn't convert it
+                throw ExceptionUtilities.Unreachable;
+            }
+
+            var stringValue = builder.ToString();
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var editor = new SyntaxEditor(root, document.Project.Solution.Workspace.Services);
+
+            editor.ReplaceNode(node, CreateUTF8StringLiteralToArrayInvocation(stringValue).WithTriviaFrom(node));
```

Comment:
This isn't really new, but removing the parameter array stuff simplified this bit, and removed 2 or 3 helper methods, and made trivia handling much simpler